File: backend/app/schemas/role.py
"""Schemas para roles y permisos."""
from uuid import UUID
from typing import Optional
from datetime import datetime

from pydantic import BaseModel, Field


# === PERMISSION ===


class PermissionResponse(BaseModel):
    id: UUID
    code: str
    display_name: str
    module: str
    description: Optional[str] = None
    sort_order: int

    class Config:
        from_attributes = True


class PermissionsByModule(BaseModel):
    """Permisos agrupados por modulo para UI."""
    module: str
    module_display: str
    permissions: list[PermissionResponse]


# === ROLE ===


class RoleCreate(BaseModel):
    name: str = Field(..., min_length=2, max_length=50)
    display_name: str = Field(..., min_length=2, max_length=100)
    description: Optional[str] = Field(None, max_length=255)
    permission_codes: list[str] = Field(
        default_factory=list,
        description="Lista de codigos de permisos a asignar",
    )


class RoleUpdate(BaseModel):
    display_name: Optional[str] = Field(None, min_length=2, max_length=100)
    description: Optional[str] = Field(None, max_length=255)
    permission_codes: Optional[list[str]] = Field(
        None, description="Lista de codigos de permisos"
    )


class RoleResponse(BaseModel):
    id: UUID
    organization_id: UUID
    name: str
    display_name: str
    description: Optional[str] = None
    is_system_role: bool
    created_at: datetime
    updated_at: datetime
    permissions: list[PermissionResponse] = []

    class Config:
        from_attributes = True


class RoleListItem(BaseModel):
    """Item para listar roles con conteos."""
    id: UUID
    name: str
    display_name: str
    description: Optional[str] = None
    is_system_role: bool
    permission_count: int
    member_count: int


# === ASSIGNMENT ===


class UserRoleAssignment(BaseModel):
    user_id: UUID
    role_id: UUID


class MyPermissionsResponse(BaseModel):
    """Permisos del usuario actual."""
    role_id: UUID
    role_name: str
    role_display_name: str
    is_admin: bool
    permissions: list[str]

File: frontend/src/types/role.ts
export interface MyPermissions {
  role_id: string;
  role_name: string;
  role_display_name: string;
  is_admin: boolean;
  permissions: string[];
  assigned_account_ids: string[];
}

export interface RoleListItem {
  id: string;
  name: string;
  display_name: string;
  description: string | null;
  is_system_role: boolean;
  permission_count: number;
  member_count: number;
}

export interface PermissionResponse {
  id: string;
  code: string;
  display_name: string;
  module: string;
  description: string | null;
  sort_order: number;
}

export interface PermissionsByModule {
  module: string;
  module_display: string;
  permissions: PermissionResponse[];
}

export interface RoleResponse {
  id: string;
  organization_id: string;
  name: string;
  display_name: string;
  description: string | null;
  is_system_role: boolean;
  created_at: string;
  updated_at: string;
  permissions: PermissionResponse[];
}

export interface RoleCreate {
  name: string;
  display_name: string;
  description?: string;
  permission_codes: string[];
}

export interface RoleUpdate {
  display_name?: string;
  description?: string;
  permission_codes?: string[];
}

export interface OrgMemberResponse {
  id: string;
  user_id: string;
  organization_id: string;
  role_id: string;
  role_name: string | null;
  role_display_name: string | null;
  joined_at: string;
  user_email: string | null;
  user_full_name: string | null;
  account_ids: string[];
  org_count: number;
}

export interface CreateUserWithMembership {
  email: string;
  full_name: string;
  role_id: string;
}
